Objects/exceptions.c

```diff
@@ -632,6 +631,290 @@ ComplexExtendsException(PyExc_BaseException, SystemExit, SystemExit,
+        PyErr_SetString(PyExc_TypeError,
+            "Expected msg followed by the nested exceptions");
+        return -1;
+    }
+    for (i = 1; i < lenargs; i++) {
+        if (!PyExceptionInstance_Check(PyTuple_GET_ITEM(args, i))) {
+            PyErr_SetString(PyExc_TypeError,
+                "nested exception must derive from BaseException");
+            return -1;
+        }
+    }
+    excs = PyTuple_GetSlice(args, 1, lenargs);
+    if (!excs) {
+        return -1;
+    }
+    for (i = 0; i < lenargs-1; i++) {
+        Py_INCREF(PyTuple_GET_ITEM(excs, i));
+    }
+    self->excs = excs;
+    self->msg = Py_NewRef(PyTuple_GET_ITEM(args, 0));
+    return 0;
+}
+
+static int
+ExceptionGroup_clear(PyExceptionGroupObject *self)
+{
+    Py_CLEAR(self->msg);
+    Py_CLEAR(self->excs);
+    return BaseException_clear((PyBaseExceptionObject *)self);
+}
+
+static void
+ExceptionGroup_dealloc(PyExceptionGroupObject *self)
+{
+    _PyObject_GC_UNTRACK(self);
+    ExceptionGroup_clear(self);
+    Py_TYPE(self)->tp_free((PyObject *)self);
+}
+
+static int
+ExceptionGroup_traverse(PyExceptionGroupObject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(self->msg);
+    Py_VISIT(self->excs);
+    return BaseException_traverse((PyBaseExceptionObject *)self, visit, arg);
+}
+
+static PyObject *
+ExceptionGroup_str(PyExceptionGroupObject *self)
+{
+    if (self->msg && PyUnicode_CheckExact(self->msg)) {
+        Py_INCREF(self->msg);
+        return self->msg;
+    }
+    else {
+        return BaseException_str((PyBaseExceptionObject *)self);
+    }
+}
+
+static PyObject* exceptiongroup_subset(PyExceptionGroupObject *orig,
+                                       PyObject *excs)
+{
+    /* Returns an ExceptionGroup with metadata from orig and exceptions from
+    the list excs.
+    excs is a subset of the exceptions in orig->excs (this is not checked).
+    */
+    PyObject *args = NULL;
+    PyObject *eg = NULL;
+    PyObject *tb = NULL;
+    PyObject *context = NULL;
+    PyObject *cause = NULL;
+
+    Py_ssize_t num_excs = PyList_GET_SIZE(excs);
+    if (num_excs < 0) {
+        return NULL;
+    }
+    else if (num_excs == 0) {
+        return Py_NewRef(Py_None);
+    }
+
+    args = PySequence_Tuple(
+        PySequence_Concat(
+            PyTuple_Pack(1, Py_NewRef(orig->msg)), PySequence_Tuple(excs)));
+
+    if (!args) {
+        goto error;
+    }
+    eg = PyObject_CallObject(PyExc_ExceptionGroup, args);
+    if (!eg) {
+        goto error;
+    }
+    tb = PyException_GetTraceback((PyObject*)orig);
+    if (tb) {
+        if (PyException_SetTraceback(eg, tb) == -1) {
+            goto error;
+        }
+    }
+    context = PyException_GetContext((PyObject*)orig);
+    if (context) {
+        PyException_SetContext(eg, context);
+    }
+    cause = PyException_GetCause((PyObject*)orig);
+    if (cause) {
+        PyException_SetCause(eg, cause);
+    }
+    return eg;
+error:
+    Py_XDECREF(eg);
+    return NULL;
```

Comment:
I believe this could result in a reference leak for the traceback object `tb` (since `PyException_GetTraceback()` returns a new ref) if the `goto error;` on line 753 is reached. Shouldn't there be a `Py_XDECREF(tb)` to decref tb when it's unable to be set to the exception?
```suggestion
    tb = PyException_GetTraceback((PyObject*)orig);
    if (tb) {
        if (PyException_SetTraceback(eg, tb) == -1) {
            goto error;
        }
    }
    context = PyException_GetContext((PyObject*)orig);
    if (context) {
        PyException_SetContext(eg, context);
    }
    cause = PyException_GetCause((PyObject*)orig);
    if (cause) {
        PyException_SetCause(eg, cause);
    }
    return eg;
error:
    Py_XDECREF(eg);
    Py_XDECREF(tb);
    return NULL;
```

Reply:
I think you're right. But should the Py_XDECREF(tb); be at the end or just before the goto error?

Right now there is no difference, but imagine some other case of "goto error" is added in the future, after tb was set successfully set on eg. Then could we not end up with double deallocation of tb (one from eg's clear)?

Reply:
> Right now there is no difference, but imagine some other case of "goto error" is added in the future, after tb was set successfully set on eg. Then could we not end up with double deallocation of tb (one from eg's clear)?

Hmm, good point. That seems reasonable to me.

Although while investigating this further, I also noticed something else regarding the interaction between `PyException_GetTraceback` and `PyException_SetTraceback` vs the other getter/setter function pairs of `Py_Exception_*` (for context and cause). For all three getters, `PyException_GetTraceback`, `PyException_GetContext`, and `PyException_GetCause`, a new reference is returned. However, only two of the setters, `PyException_SetContext` and `PyException_SetCause`, actually "steal" a reference, using `Py_XSETREF` directly (e.g. resulting in +1 references for the `context` object from `PyException_GetContext` and -1 from `PyException_SetContext`).

In `PyException_SetTraceback`, there seems to be a net 0 difference though, since in its underlying implementation function (`BaseException_set_tb`), there is a `Py_Incref` just before the `Py_XSETREF`, unlike the other two setters. So overall, for the traceback, we get +1 references to the `tb` object from `PyException_GetTraceback`, +1 from the `Py_Incref`, and -1 from `Py_XSETREF`. This results in a net difference of +1 references for the tb object (within `exceptiongroup_subset`), so don't we want to have `Py_XDecref(tb)` at the before returning from `exceptiongroup_subset`, regardless of whether `PyException_SetTraceback` succeeds?

Note that the C-API isn't really my area of expertise, so I could just be misunderstanding something here. But I wanted to mention it anyways in case it was overlooked and for my own clarification.

Reply:
Yes, I think you're right about this as well. It should be something equivalent to this then:

```
if (tb) {
        if (PyException_SetTraceback(eg, tb) == -1) {
            Py_XDECREF(tb);
            goto error;
        } else {
            Py_XDECREF(tb);
        }
    }
```